File: initial/sorting/merge-sort.js
function mergeSort(arr) {
	if (arr.length <= 1) return arr;

	const middle = Math.floor(arr.length / 2);
	const left = mergeSort(arr.slice(0, middle));
	const right = mergeSort(arr.slice(middle, arr.length));

	return merge(left, right);
}

console.log("mergeSort", mergeSort([3, 5, 7, 1, 9, 2, 100, 99, 77, 88]));

// O(n + m) time complexity and O(n + m) time complexity
function merge(arr1, arr2) {
	let i = 0;
	let j = 0;
	let res = [];

	while (i < arr1.length && j < arr2.length) {
		if (arr1[i] > arr2[j]) {
			res.push(arr2[j]);
			j++;
		} else {
			res.push(arr1[i]);
			i++;
		}
	}

	if (i < arr1.length) {
		res = res.concat(arr1.slice(i, arr1.length));
	}

	if (j < arr2.length) {
		res = res.concat(arr2.slice(j, arr2.length));
	}

	return res;
}

// console.log(merge([1, 3, 5, 6, 7, 11], [1, 2, 4, 8, 10]));

File: initial/sorting/merge-sort.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func main() {
	start := time.Now();
	// defer func() {
	// 	fmt.Printf("mergeSort() for 7 items took %v\n", time.Since(start))
	// }()
	// fmt.Println("mergeSort()", mergeSort([]int{1, 10, 50, 2, 14, 99, 100}))

	defer func() {
		fmt.Printf("mergeSort() for 1k items took %v\n", time.Since(start))
	}()
	param := []int{}
	// randomize 10k unsorted items for param slice
	for i := 0; i < 10000; i++ {
		param = append(param, rand.Intn(10000))
	}

	fmt.Println("mergeSort()", mergeSort(param))
}

func mergeSort(arr []int) []int {

	if len(arr) <= 1 {
		return arr
	}

	mid := len(arr) / 2
	left := mergeSort(arr[:mid])
	right := mergeSort(arr[mid:])

	return merge(left, right)
}

func merge(arr1 []int, arr2 []int) []int {
	i := 0
	j := 0
	res := []int{}

	for i < len(arr1) && j < len(arr2) {
		if arr1[i] < arr2[j] {
			res = append(res, arr1[i])
			i++
		} else {
			res = append(res, arr2[j])
			j++
		}
	}

	if i < len(arr1) {
		res = append(res, arr1[i:]...)
	}

	if j < len(arr2) {
		res = append(res, arr2[j:]...)
	}

	return res;
}